exp/operator/controllers/coreprovider_controller.go

```diff
@@ -19,31 +19,80 @@ package controllers
 import (
 	"context"
 
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	operatorv1 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/patch"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller"
-
-	operatorv1alpha4 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
 )
 
 // CoreProviderReconciler reconciles a CoreProvider object.
 type CoreProviderReconciler struct {
 	Client client.Client
 }
 
+func (r *CoreProviderReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&operatorv1.CoreProvider{}).
+		WithOptions(options).
+		Complete(r)
+}
+
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operator.cluster.x-k8s.io,resources=coreproviders/status,verbs=get;update;patch
 
-func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	log := ctrl.LoggerFrom(ctx)
-	log.V(4).Info("Reconcile CoreProvider")
+func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	coreProvider := &operatorv1.CoreProvider{}
 
-	return ctrl.Result{}, nil
+	if err := r.Client.Get(ctx, req.NamespacedName, coreProvider); err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, err
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(coreProvider, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		if err := patchHelper.Patch(ctx, coreProvider); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	// Ignore deleted CoreProvider, this can happen when foregroundDeletion
+	// is enabled
```

Comment:
Maybe expand to explain why we don't need to clean up? Owner references will cause the GC to clean up all the resources created by the `CoreProvider` right?

Reply:
yeah, that's how I got it. https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion

Reply:
There is an ongoing discussion about foreground deletion in https://github.com/kubernetes-sigs/cluster-api/issues/4085, and I think we should wait for this in order to address the problem consistently across the board.
Also, this relies on owner chain, and I'm not sure we are already at a stage where the chain of objects managed by each provider is yet fully defined.